### ISSUE TITLE
Extraction filter with time format function not working (0.8.2)

### ISSUE THREAD

**Issue by ryryguy** [apache/incubator-druid#2652]
_(maybe user error and not a defect? But I posted on druid-dev and got no response, so I'm going to assume defect.)_

I can't get an extraction filter, which uses a time format extraction function, to return any results. 

I'm using the batch ingestion wikipedia example - the six rows of data included at examples/indexing/wikipedia_data.json.  (Data also appears here.) 

Also, I'm running only a historical and broker node (because that's how my integration tests are set up to run). I'm querying the historical node directly. Druid version 0.8.2

Using the time format extraction function as a dimension spec works just fine:

Query:

```
{
  "dataSource": "druid-client-test",
  "dimensions": [
    "continent",
    {
      "dimension": "__time",
      "extractionFn": { "format": "HH", "type": "timeFormat"},
      "outputName": "hod",
      "type": "extraction"
    }
  ],
  "granularity": "all",
  "aggregations": [{"name": "count_it", "type": "count"}],
  "intervals": ["2013-08-31T01:02:33.000Z/2013-09-01T12:41:27.000Z"],
  "queryType": "groupBy"
}
```

Result:

```
[ {
  "version" : "v1",
  "timestamp" : "2013-08-31T01:02:33.000Z",
  "event" : {
    "continent" : "Asia",
    "count_it" : 1,
    "hod" : "07"
  }
}, {
  "version" : "v1",
  "timestamp" : "2013-08-31T01:02:33.000Z",
  "event" : {
    "continent" : "Asia",
    "count_it" : 1,
    "hod" : "11"
  }
}, {
  "version" : "v1",
  "timestamp" : "2013-08-31T01:02:33.000Z",
  "event" : {
    "continent" : "Asia",
    "count_it" : 1,
    "hod" : "12"
  }
}, {
  "version" : "v1",
  "timestamp" : "2013-08-31T01:02:33.000Z",
  "event" : {
    "continent" : "Australia",
    "count_it" : 1,
    "hod" : "03"
  }
}, {
  "version" : "v1",
  "timestamp" : "2013-08-31T01:02:33.000Z",
  "event" : {
    "continent" : "North America",
    "count_it" : 1,
    "hod" : "01"
  }
} ]
```

But if I add a filter to the query that uses the exact same extraction function, I get no results:

Query:

```
{
  "dataSource": "druid-client-test",
  "dimensions": [
    "continent",
    {
      "dimension": "__time",
      "extractionFn": { "format": "HH", "type": "timeFormat"},
      "outputName": "hod",
      "type": "extraction"
    }
  ],
  "granularity": "all",
  "aggregations": [{"name": "count_it", "type": "count"}],
  "intervals": ["2013-08-31T01:02:33.000Z/2013-09-01T12:41:27.000Z"],
  "filter": {
     "dimension": "__time",
     "value": "01",
     "extractionFn": { "format": "HH", "type": "timeFormat" },
     "type": "extraction"
  },
            "queryType": "groupBy"
}
```

Result:

`[ ]`

**Comment by fjy** [comment]
@ryryguy it takes us a few days to get around to community issues

**Comment by fjy** [comment]
@ryryguy I thikn what you are trying to do is just return hourly results. Use granularity "hour" and remove the extractoinFn. You might also just want to use https://github.com/implydata/plyql if you are more comfortable with SQL.

**Comment by ryryguy** [comment]
Thanks Fangjin!

But no, actually I am not trying to get hourly results. I am trying to get results grouped by **hour of day**.

The example above based on the Wikipedia example data set may not be clear, because that data set only contains data for a single day. Consider this alternate example data set: 

```
{"timestamp": "2013-08-29T01:02:33Z", "page": "Gypsy Danger", "added": 57, "deleted": 200, "delta": -143}
{"timestamp": "2013-08-29T12:32:45Z", "page": "Striker Eureka", "added": 459, "deleted": 129, "delta": 330}
{"timestamp": "2013-08-30T01:11:21Z", "page": "Cherno Alpha", "added": 123, "deleted": 12, "delta": 111}
{"timestamp": "2013-08-31T11:58:39Z", "page": "Crimson Typhoon", "added": 905, "deleted": 5, "delta": 900}
{"timestamp": "2013-08-31T12:41:27Z", "page": "Coyote Tango", "added": 1, "deleted": 10, "delta": -9}
```

I've altered the timestamps a bit. Now, if I do a query similar to the first one in my original example but omitting the "continent" dimension but using the time extraction function, I should get three buckets, for the following values of "hod":

"01" (count = 2: "Gypsy Danger" and "Cherno Alpha")
"11" (count = 1: "Crimson Typhoon")
"12" (count = 2: "Striker Eureka", "Coyote Tango")

(An "hour" granularity query would return five buckets, right? That is not what I'm trying to get.)

So that works as expected for the query itself. But the same time extraction function in the filter, where say I might want to get results where hour of day is only "01", returns no results.

**Comment by ryryguy** [comment]
So the postgresql equivalent to what I'm trying to do:

```
SELECT COUNT(value),
  EXTRACT(hour FROM timestamp) AS hod
    FROM table WHERE hod = '01'
      GROUP BY hod
```

Granularity = "hour" would, I think, be equivalent to using postgresql DATE_TRUNC instead of EXTRACT. `DATE_TRUNC('hour', TIMESTAMP '2001-02-16 20:38:40')` returns "2001-02-16 20:00:00". `EXTRACT('hour', TIMESTAMP '2001-02-16 20:38:40')` returns "20".

**Comment by vogievetsky** [comment]
I was looking into this to answer @ryryguy question and I think there might be a bug in Druid.

For me the query:

```json
{
  "queryType": "timeseries",
  "dataSource": "wikipedia",
  "intervals": "2015-09-12/2015-09-13",
  "granularity": "hour",
  "context": {
    "useCache": false,
    "populateCache": false
  },
  "filter": {
    "type": "extraction",
    "dimension": "__time",
    "value": "3",
    "extractionFn": {
      "type": "timeFormat",
      "format": "H",
      "timeZone": "Etc/UTC",
      "locale": "en-US"
    }
  },
  "aggregations": [{ "name": "RowCount", "type": "count" }]
}
```

Returns:

```json
[ {
  "timestamp" : "2015-09-12T00:46:00.000Z",
  "result" : {
    "RowCount" : 0
  }
} ]
```

But the expected value is `8072` as you can see in the query:

```json
{
  "queryType": "topN",
  "dataSource": "wikipedia",
  "intervals": "2015-09-12/2015-09-13",
  "granularity": "all",
  "context": {
    "useCache": false,
    "populateCache": false
  },
  "aggregations": [{ "name": "RowCount", "type": "count" }],
  "dimension": {
    "type": "extraction",
    "dimension": "__time",
    "outputName": "HourOfDay",
    "extractionFn": {
      "type": "timeFormat",
      "format": "H",
      "timeZone": "Etc/UTC",
      "locale": "en-US"
    }
  },
  "metric": {
    "type": "alphaNumeric"
  },
  "threshold": 1000
}
```

*Correctly* returns:

```json
[ {
  "timestamp" : "2015-09-12T00:46:00.000Z",
  "result" : [ {
    "RowCount" : 2644,
    "HourOfDay" : "0"
  }, {
    "RowCount" : 11329,
    "HourOfDay" : "1"
  }, {
    "RowCount" : 10946,
    "HourOfDay" : "2"
  }, {
    "RowCount" : 8072,
    "HourOfDay" : "3"
  }, {
    "RowCount" : 8159,
    "HourOfDay" : "4"
  }, {
    "RowCount" : 12540,
    "HourOfDay" : "5"
  }, {
    "RowCount" : 21126,
    "HourOfDay" : "6"
  }, {
    "RowCount" : 22261,
    "HourOfDay" : "7"
  }, {
    "RowCount" : 16093,
    "HourOfDay" : "8"
  }, {
    "RowCount" : 16914,
    "HourOfDay" : "9"
  }, {
    "RowCount" : 17799,
    "HourOfDay" : "10"
  }, {
    "RowCount" : 16089,
    "HourOfDay" : "11"
  }, {
    "RowCount" : 16910,
    "HourOfDay" : "12"
  }, {
    "RowCount" : 20537,
    "HourOfDay" : "13"
  }, {
    "RowCount" : 18556,
    "HourOfDay" : "14"
  }, {
    "RowCount" : 19488,
    "HourOfDay" : "15"
  }, {
    "RowCount" : 19407,
    "HourOfDay" : "16"
  }, {
    "RowCount" : 22814,
    "HourOfDay" : "17"
  }, {
    "RowCount" : 21506,
    "HourOfDay" : "18"
  }, {
    "RowCount" : 19970,
    "HourOfDay" : "19"
  }, {
    "RowCount" : 18159,
    "HourOfDay" : "20"
  }, {
    "RowCount" : 17532,
    "HourOfDay" : "21"
  }, {
    "RowCount" : 15784,
    "HourOfDay" : "22"
  }, {
    "RowCount" : 14684,
    "HourOfDay" : "23"
  } ]
} ]
```

This is running agains the Plywood test docker. To reproduce it simply setup the Docker container in the "Druid" section of: https://github.com/implydata/plywood/tree/master/data and run the queries posted with.

**Comment by jon-wei** [comment]
Think I found the cause:

In ExtractionFilter:
```
  @Override
  public ImmutableBitmap getBitmapIndex(BitmapIndexSelector selector)
  {
    final List<Filter> filters = makeFilters(selector);
    if (filters.isEmpty()) {
      return selector.getBitmapFactory().makeEmptyImmutableBitmap();
    }
    ...
```
```
  private List<Filter> makeFilters(BitmapIndexSelector selector)
  {
    Indexed<String> allDimVals = selector.getDimensionValues(dimension);
    final List<Filter> filters = Lists.newArrayList();
    if (allDimVals == null) {
      allDimVals = new Indexed<String>()
      {
        @Override
        public Iterator<String> iterator()
        {
          return null;
        }

        @Override
        public Class<? extends String> getClazz()
        {
          return null;
        }

        @Override
        public int size() { return 1; }

        @Override
        public String get(int index) { return null;}

        @Override
        public int indexOf(String value)
        {
          return 0;
        }
      };
    }

    for (int i = 0; i < allDimVals.size(); i++) {
      String dimVal = allDimVals.get(i);
      if (value.equals(Strings.nullToEmpty(fn.apply(dimVal)))) {
        filters.add(new SelectorFilter(dimension, dimVal));
      }
    }

    return filters;
  }
```

```
  @Override
  public Indexed<String> getDimensionValues(String dimension)
  {
    final Column columnDesc = index.getColumn(dimension);
    if (columnDesc == null || !columnDesc.getCapabilities().isDictionaryEncoded()) {
      return null;
    }
    ...
```

The filter tries to access the column's dictionary to determine which values (those that map to the targeted extracted value) need to be matched on, but the time column does not have one, so makeFilters eventually returns an empty list -> empty result set

**Comment by vogievetsky** [comment]
I can confirm that everything works fine for non-primary time dimensions (as in not __time).

**Comment by fjy** [comment]
@jon-wei can u add a fix and a test?

**Comment by jon-wei** [comment]
@fjy I have a patch that allows this case to work:

https://github.com/druid-io/druid/pull/2621/commits/f9b4da5f5d2d48a40e03ac7aa6c5e4438907566b

However, it depends on some interfaces introduced in my long/float dimension types PR for performing a filtered scan on columns without bitmap indexes.

Do you think we should implement a fix with a column scan implementation for time column only for now, or wait for that PR to be brought in (I expect that PR will take a while to go through review)?

**Comment by ryryguy** [comment]
Jon how much worse is the performance of the column scan vs the filtered scan? 

(How much worse is it than using the existing filter functionality targeting an actual time dimension rather than "__time"? Because I'm probably going to try doing that as a work around to satisfy our use case. We may not be able to update our Druid to a version containing the fix right away. Though if the column scan implementation is equivalent to or better than targeting an actual dimension, we might try applying it to a fork of 0.8.2.)

**Comment by jon-wei** [comment]
@fjy I'm currently taking a look at what's involved in creating a patch that just allows these filters to work on __time using full column scan, if it's not a huge patch with big overlap with the dim types patch, I can separate that out.

@ryryguy Do you mean, in the alternative, ingesting a second "time" dimension with String timestamps?

I'm not sure about performance difference, I haven't benchmarked the two cases, but it probably depends on the cardinality of the timestamps. 

It would come down to the difference between scanning all the rows in the __time column (stored as longs) and applying the filter vs. iterating through all the dictionary values for the second "time" column (standard druid string dimension) and performing a union of the filtered row bitsets.

**Comment by jon-wei** [comment]
@fjy @ryryguy 

I've opened a PR that will allow filtering on the __time column, some benchmark results are included there:

https://github.com/druid-io/druid/issues/2673

**Comment by ryryguy** [comment]
That looks great. Thanks for the benchmarks too!

So this PR is not dependent on the "dim types" patch you mentioned? I could theoretically apply it to druid 0.8.2? (Not sure if I actually will, we're tentatively planning to upgrade to 0.9.0 when it's released - did this make it in time for that release?)

**Comment by fjy** [comment]
@ryryguy not sure if this will make it into 0.9.0 as it is very late in the release cycle, but it'll be part of 0.9.0-iap if you need a downloadable version, otherwise it'll definitely be a part of 0.9.1

**Comment by vogievetsky** [comment]
This was fixed in 0.9.1
